core/rpc/src/rpc_impl.rs

```diff
@@ -135,33 +133,61 @@ where
     }
 
     fn register_addresses(&self, normal_addresses: Vec<String>) -> RpcResult<Vec<H160>> {
-        let mut ret = Vec::new();
-        let mut batch = rpc_try!(self.store.batch());
-
-        // Todo: refactor this.
-        for addr in normal_addresses.iter() {
-            let script = address_to_script(Address::from_str(addr).unwrap().payload());
-            let script_hash = blake2b_160(script.as_slice());
-            let key = add_prefix(
-                *SCRIPT_HASH_EXT_PREFIX,
-                script_hash::Key::ScriptHash(script_hash).into_vec(),
-            );
-
-            rpc_try!(batch.put_kv(key, script_hash::Value::Script(&script)));
-            ret.push(H160(script_hash));
-        }
-
-        rpc_try!(batch.commit());
+        log::debug!("register addresses {:?}", normal_addresses);
+        self.inner_register_addresses(normal_addresses)
+            .map_err(|e| Error::invalid_params(e.to_string()))
+    }
 
-        Ok(ret)
+    fn get_generic_transaction(&self, tx_hash: H256) -> RpcResult<GenericTransactionWithStatus> {
+        log::debug!("get generic transaction tx hash {:?}", tx_hash);
+        let tx = rpc_try!(block_on!(self, get_transactions, vec![tx_hash]))
+            .get(0)
+            .cloned()
+            .unwrap()
+            .unwrap();
+        let tx_hash = tx.transaction.hash;
+        let tx_status = tx.tx_status.status;
+
+        self.inner_get_generic_transaction(tx.transaction.inner.into(), tx_hash, tx_status)
+            .map_err(|e| Error::invalid_params(e.to_string()))
     }
 
-    fn scan_deposit(&self, payload: ScanBlockPayload) -> RpcResult<ScanBlockResponse> {
-        log::debug!("query charge payload {:?}", payload);
-        self.inner_scan_deposit(
-            payload.block_number,
-            payload.udt_hash.clone(),
-            payload.idents,
+    fn get_generic_block(&self, payload: GetGenericBlockPayload) -> RpcResult<GenericBlock> {
+        let current_number = **CURRENT_BLOCK_NUMBER.load();
+        let use_hex_format = **USE_HEX_FORMAT.load();
+
+        let block = if payload.block_num.is_some() {
+            let num = payload.block_num.unwrap();
+            if num > current_number {
+                return Err(Error::invalid_params("invalid block number"));
+            }
+
+            let resp = rpc_try!(block_on!(self, get_block_by_number, num, use_hex_format)).unwrap();
+            if let Some(hash) = payload.block_hash {
+                if resp.header.hash != hash {
+                    return Err(Error::invalid_params("block number and hash mismatch"));
+                }
+            }
+            resp
+        } else if payload.block_hash.is_some() && payload.block_num.is_none() {
+            let hash = payload.block_hash.unwrap();
+            rpc_try!(block_on!(self, get_block, hash, use_hex_format)).unwrap()
+        } else {
```

Comment:
if block_hash and block_num are both none, query the latest block, not return error